### ISSUE TITLE
Spell Checking

### ISSUE THREAD

**Issue by rkusa** [zed-industries/zed#5305]
**Is your feature request related to a problem? Please describe.**
I regularely make a lot of typos in code comments, docs, but also in variable names. It is always annoying if pull-requests get post-poned just because of typos reviewers found.

**Describe the solution you'd like**
I find it tremendously helpful if Zed could spell checks my code. I personally rely a lot on this in e.g.:
- Sublime Text: https://www.sublimetext.com/docs/spell_checking.html, and
- VSCode: https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker.

Things the spell checker might check:
- comments,
- doc comments,
- strings,
- segments of a variable name (eg. in `get_name`/`GetName`/`getName`, it would check `get` and `name`).

Additional useful features:
- The possibility to add words to a system-wide dictionary.
- The possibility to add words to a project-specific dictionary.
- The possibility to check multiple languages at the same time (e.g. for non-English code, variables are often in English, while strings and comments might contain text in another language).

I'd find it especially neat if the spell check would use native system APIs so that I can expect consistent behaviour between the editor and other apps on my system.

**Screenshots**

Sublime Text:

![image](https://user-images.githubusercontent.com/409021/178147096-e46229ee-8d08-47f7-9506-caec62b20347.png)

VSCode Code Spell Checker extension:

![image](https://user-images.githubusercontent.com/409021/178147007-b39a2a75-4a55-4bfc-9277-db91b2b8556a.png)

**Comment by jaydenseric** [comment]
I just tried setting up Zed today and this was the showstopper issue that made me go back to VS Code. Spell check is a critical concern of an editor; it's generally too complex to enforce correct spelling cross-platform via CLI and CI with any degree of confidence and you may be contributing spelling errors to projects with tooling and CI not in your control. I am hopelessly dependent on spellcheck when writing hundreds of words worth of comments and docs per day, or when naming things like types, functions, and variables.

VS Code spell check is quite poor because it's via an extension (https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) and not the native macOS spellcheck system, so the spell check UI is inconsistent with other apps and the learned words and not in sync with every other app on your system. But at least it has a solution.

There are non VS Code editors out there that have really nice native OS based spell check, so it's possible.

**Comment by SpyrosMourelatos** [comment]
I would prefer LanguageTool to Grammarly as it FOSS

**Comment by JosephTLyons** [comment]
This feature feels like a rather large one to add to Zed, when considering all of the things you should be able to do with spell check. What does a good first pass look like? What would be the bare minimum needed to ship something useable?

Some unknowns:

- Is there a Rust crate out there suitable for spell checking in the context of code?
- Would we have some sort of setting for which file types to spell check, or, maybe a setting for which file types to not spell check.. or both?  I think VS Code's spell check has these settings.
- How do we want to surface the spelling errors? Is it just another red squiggly line and do we have a way to suggest a fix?  I feel like for a first pass, we could simply display the potential matches in the hover, and a future follow-up could add the possibility to accept a suggestion for a typo.
- VS Code provides spell checking for multiple languages, do we just ship English first? Maybe we'll be lucky and have a crate that can handle multiple languages.

Bonus points: Spell checking in Zed's chat editor.

A long-term stretch goal might be to somehow leverage the supported AI models in Zed to fill suggestions for mispelled words.

**Comment by bajtos** [comment]
> How do we want to surface the spelling errors? Is it just another red squiggly line and do we have a way to suggest a fix? 

In VS Code, the Grammarly extension integrates with the "suggested fix" feature provided by language servers. I can use the same keyboard shortcut to fix ESlint violations and spelling/grammar mistakes.

**Comment by jansol** [comment]
[Vale](https://vale.sh/) is an offline rule-based "prose linter" (spelling & style checker) with an official LSP implementation. It is also code-aware so it can check code comments and won't get confused by markdown or HTML. Seems like a great fit?

**Comment by arthur-st** [comment]
Testing Vale this week, and it works quite well on the CLI level. The configuration might be a bit tricky to set up, but should be fine sailing afterwards.

**Comment by phaynes** [comment]
> ## Future AI ideas
> A distant future goal might be to somehow leverage the supported AI models in Zed to fill suggestions for mispelled words, if the suggestions provided by some crate aren't the greatest.

I frequently switch among Zed, ChatGPT, Gemini, etc., for Markdown editing, focusing on three key tasks:

1. *Proofing*: My standard response settings are tailored for specific styles, as exemplified below. I need the proofreading to adhere to a distinct style based on the context—basic questions, formal responses, and technical discourse—and to be delivered in Markdown format using British English spelling.

For example, part of my chatGPT response template is -
 "Basic questions should be concise and direct, emulating the style of the Australian Daily Telegraph, without colloquialisms. Formal responses should be brief and direct, utilizing a military writing style. Technical responses should cater to a postgraduate audience and remain technical."
Ideally, I'd like the option to easily select the level of proofing required right after highlighting the text I've just written. i.e a "quick proof", followed by a more reasoned proof, where I give the AI's more context to assist in their feedback.

2. *Argument Improvement & Revision*: I alternate between AI tools to refine and enhance my work. Direct integration of such feedback within Zed would streamline my editing process.

3. *Post-Production*: The final adjustments include ensuring APA v7 compliance for references, adding document headers, and formatting the output to specific standards (PDF/HTML or voice script). 

Incorporating these features into Zed would significantly enhance my editing workflow, minimizing the need to switch between different tools for each task, and negate the need for inline spelling checks

**Comment by brandondrew** [comment]
> I would prefer LanguageTool to Grammarly as it FOSS and supports more languages

Yeah, I've never been overly impressed with the feedback I get from Grammarly.

**Comment by levlaz** [comment]
I'm trying to use zed to write docs in MD and MDX. The lack of a spell checker is really painful.

**Comment by Retkid** [comment]
> > ## Future AI ideas
> > A distant future goal might be to somehow leverage the supported AI models in Zed to fill suggestions for mispelled words, if the suggestions provided by some crate aren't the greatest.
> 
> I frequently switch among Zed, ChatGPT, Gemini, etc., for Markdown editing, focusing on three key tasks:
> 
>     1. _Proofing_: My standard response settings are tailored for specific styles, as exemplified below. I need the proofreading to adhere to a distinct style based on the context—basic questions, formal responses, and technical discourse—and to be delivered in Markdown format using British English spelling.
> 
> 
> For example, part of my chatGPT response template is - "Basic questions should be concise and direct, emulating the style of the Australian Daily Telegraph, without colloquialisms. Formal responses should be brief and direct, utilizing a military writing style. Technical responses should cater to a postgraduate audience and remain technical." Ideally, I'd like the option to easily select the level of proofing required right after highlighting the text I've just written. i.e a "quick proof", followed by a more reasoned proof, where I give the AI's more context to assist in their feedback.
> 
>     2. _Argument Improvement & Revision_: I alternate between AI tools to refine and enhance my work. Direct integration of such feedback within Zed would streamline my editing process.
> 
>     3. _Post-Production_: The final adjustments include ensuring APA v7 compliance for references, adding document headers, and formatting the output to specific standards (PDF/HTML or voice script).
> 
> 
> Incorporating these features into Zed would significantly enhance my editing workflow, minimizing the need to switch between different tools for each task, and negate the need for inline spelling checks

This is very common for me as well. As someone who struggles with formatting and grammar, I use chatGPT and Gemini to help me spot issues. Though, I don't usually use them for their generative capabilities. A powerful spell and grammar checker like Grammarly would greatly assist this. I don't think, however, you need to integrate any AI into the codebase.

**Comment by phaynes** [comment]
Hi, 

I have taken a first stab at creating a configurable full spell checker / grammar checker and proofing engine that **integrates to Zed** - example key bindings included, uses the OpenAI and Anthropic API's. 

The [markdown proofing engine](https://github.com/phaynes/markdown-processing/tree/main/markdown-proofer) is here, and I am finalising an initial baseline to part of a general publication engine from text - although I am starting with research papers. 

Any and all feedback would be greatly appreciated. This is genuinely a first drop of the approach.

Philip

**Comment by florinpatrascu** [comment]
I truly appreciate the effort the team and contributors put into bringing this feature to life, but I have to ask: is there any chance of getting a spellchecker for Zed that doesn't rely on a remote service or require a local GPU?

**Comment by jansol** [comment]
> I truly appreciate the effort the team and contributors put into bringing this feature to life, but I have to ask: is there any chance of getting a spellchecker for Zed that doesn't rely on a remote service or require a local GPU?

The Vale extension does *exactly* this. It relies on a local dictionary (defined with plain text files). Unfortunately it currently advertises support for Markdown files, and the language server was crashing a lot when used from zed last time I tried. Nobody really knew *why*, though.

**Comment by phaynes** [comment]
The short answer is yes, if you are happy to run a local LLM. I performed a smoke test of the concept using Llama-3.1-8b-Instant on [groq](https://groq.com), and this seems to be viable.

Would a solution that uses Llama-3.1-8b running on something like Ollama locally work for you, or running Llama inside a Docker container?

The engine above is designed to support multiple AI engines (via the -ai flag), but it needs checking and would likely require modifications to the configurable prompts.

**Some more explanation:**

The goal of the above engine is to not only consider spell checking for different international languages but also complex grammar checking and proofing against a range of criteria required for publication-grade technical documentation. 

Thus, spell checking the following should be fine:
- **Definition:** Acids are hydrogen-containing compounds that dissociate in water to give H\(^+\) ions.
- **Equation:** \(\text{HCl(aq)} \rightarrow \text{H}^+(\text{aq}) + \text{Cl}^-(\text{aq})\)

As should the following prompt:
```code
Without providing comments, proof the following text for spelling and grammar, using British English, active voice, markdown, academic style, and do not change BibTeX references.
```

Furthermore, spell and grammar checking within a software IDE requires an understanding of code and data formats such as JSON to avoid the ongoing frustration of not being able to closely integrate code and data across the full large-scale software engineering lifecycle.

While it is conceivable these considerations could be met otherwise, with the advent of high-quality open-source LLMs, I believe it would be cost prohibitive to do so.

**Comment by jansol** [comment]
> While it is conceivable these considerations could be met otherwise, with the advent of high-quality open-source LLMs, I believe it would be cost prohibitive to do so.

An understandable concern, this has only been done without LLMs for decades after all. Faster and more deterministically than a LLM ever will, too.

It is truly fascinating how quickly people have discarded the very idea of even trying to do anything without a LLM.

**Comment by florinpatrascu** [comment]
> Would a solution that uses Llama-3.1-8b running on something like Ollama locally work for you, or running Llama inside a Docker container?

As a non-native English speaker, I find a standard spellchecker incredibly helpful. If I have to run any LLM in the background or use Docker for such tasks, it would diminish my enjoyment of coding. Additionally, I do not consider Docker an essential tool for my workstation — servers or similarly related tasks/needs, perhaps, but not for my personal setup. This is just my personal preference, of course.

**Comment by phaynes** [comment]
> An understandable concern, this has only been done without LLMs for decades after all. Faster and more deterministically than a LLM ever will, too.
> 
> It is truly fascinating how quickly people have discarded the very idea of even trying to do anything without a LLM.

The x86 assembly source code for my 1984 scientific word processor was indeed fast but is unfortunately lost to history. However, your assertion that "people have quickly discarded the idea of doing anything without an LLM" is incorrect and underestimates the complexity of the problem I have outlined.

Large companies like Grammarly exist for a reason: grammar checking, citations, and writing to a style are challenging problems, particularly for any spoken language. This is before considering how these technologies must be closely integrated within software environments and optimized against specific criteria. These functions are indeed well-suited for LLMs.

As I stated, my goal is to have the capability to produce publication-grade documentation from the source. This capability is a key mechanism for whole lifecycle integration in large software-intensive programs.

If this is not within your scope, that is okay too.

**Comment by phaynes** [comment]
> As a non-native English speaker, I find a standard spellchecker incredibly helpful. If I have to run any LLM in the background or use Docker for such tasks, it would diminish my enjoyment of coding. Additionally, I do not consider Docker an essential tool for my workstation — servers or similarly related tasks/needs, perhaps, but not for my personal setup. This is just my personal preference, of course.
I’m glad I asked! Would the following option be useful instead: a simpler, dictionary-based English spell checker that assumes the entire line is text? It would handle only basic spelling errors within the text block, leaving grammar and more sophisticated proofing for a different stage.

**Comment by Altair-Bueno** [comment]
Almost every spellchecker out there runs on plain old dictionary checks. Think of your iPhone or Android keyboards. Your Mac. Every browser you've used. They all run with almost basic spell checking and thats alright. lenght, heigth, gramar, enginer, ...

Over complicating spell checking with LLMs brings no tangible benefits to the table but rather disadvantages such as higher resource usage, slower feedback loops and non deterministic results.

**Comment by phaynes** [comment]
> > Would a solution that uses Llama-3.1-8b running on something like Ollama locally work for you, or running Llama inside a Docker container?
> 
> As a non-native English speaker, I find a standard spellchecker incredibly helpful. If I have to run any LLM in the background or use Docker for such tasks, it would diminish my enjoyment of coding. Additionally, I do not consider Docker an essential tool for my workstation — servers or similarly related tasks/needs, perhaps, but not for my personal setup. This is just my personal preference, of course.

@florinpatrascu actually, I have a different idea that I would appreciate your input on.

**Question 1:** As a non-native English speaker, would the ability to write initial comments in your native language and then translate them to English be beneficial?

**Context**:

“Spell checking” can mean various things: from simple autocorrect in mobile messaging to spell checking in text, basic spell checking in code, and the more rigorous spell checking required in formal system documents, typically for larger software systems. To avoid the complexity of writing parsers for every language and context, robustly identifying written words in text, and enabling sentence-level contextual spell checking:

**Question 2:** Would having a small embedded LLM periodically run on your local machine as needed to profile your code statistically, simplify spell checking, and incorporate custom rules (e.g., ignoring LaTeX/references), be acceptable?

**Comment by brandondrew** [comment]
> Almost every spellchecker out there runs on plain old dictionary checks. Think of your iPhone or Android keyboards. Your Mac. Every browser you've used. They all run with almost basic spell checking and thats alright. lenght, heigth, gramar, enginer, ...
> 
> Over complicating spell checking with LLMs brings no tangible benefits to the table but rather disadvantages such as higher resource usage, slower feedback loops and non deterministic results. It's a _Code editor_, not MS Word. I ain't writing Harry Potter books on it.

Your 2 points are at odds with each other.  First you want a simple spell checker that "runs on plain old dictionary checks".  Next you point out that Zed is a _code editor_.  

While I can see why someone would choose Zed to write a novel in--it's a pretty awesome editor--you're exactly right that Zed is primarily aimed at editing code.  This very fact makes the spell checker much more complicated than you seem to realize.  For example, do you want your spell checker complaining about every token in your code that doesn't match a word in the dictionary?  That's what you would get with nothing but "plain old dictionary checks".

**Comment by Altair-Bueno** [comment]
> do you want your spell checker complaining about every token in your code that doesn't match a word in the dictionary

Yes. Exactly. That's a great behavior. Even better if I can add my own custom dictionaries. For non-native English speakers, such as florin and I, that's all we need. And even English speakers would benefit from it.

However, it needs to be mindful of case style (e.g CollectorBuilder should be transformed into collector and builder and check separately).

**Comment by phaynes** [comment]
> > do you want your spell checker complaining about every token in your code that doesn't match a word in the dictionary
> 
> Yes. Exactly. That's a great behaviour. 

No, that is terrible behavior and will never work.

I sit in the camp of writing code, yet I somehow find myself writing 30-40K+ high-quality words per year across the full SDLC.

My guess is that it is important Zed correctly supports international work. The spell checker can’t disrupt global build systems when people from the Commonwealth insist on using British English, and US teams dig their heels in, insisting on American English. Also, we all know how much the French love having their code comments translated to English, and that Japanese teams prefer writing and reading code in a language other than US English.

For this reason, I think the spell checker needs to work in batch mode with common parameters. In this way, code comments are saved and then checked into a common baseline, then translated to the programmer’s spoken language of choice on the fly when opening a file, permitting editing of the file in the person’s native language. When checking the code back in, the written words are transformed back into the baseline language. This is just solving a more complicated version of the spaces/tabs and curly braces wars.

Additionally, the near-ubiquitous use of code within comments pretty much requires some sort of smart parameterization so only written words are checked. The spell checker needs to work not only in mainline code but with config and data files of every flavor. Again, this will need to be tailored to the circumstances. I can’t see how this can easily be done without the use of LLMs.

Now, I definitely think an embedded mode would be super useful—and indeed essential when development environments are not internet-connected. Given that LLMs can also generate code, an acceptable compromise could be for the LLM to pre-assess code locally, allowing simplified spell checkers to be generated on the fly and permit more lightweight checking.

Anyway, these are my thoughts and relatively straightforward enhancements based on the work I have done above.

Philip

**Comment by brandondrew** [comment]
> > do you want your spell checker complaining about every token in your code that doesn't match a word in the dictionary
> 
> Yes. Exactly. That's a great behavior. Even better if I can add my own custom dictionaries. For non-native English speakers, such as florin and I, that's all we need. And even English speakers would benefit from it.
> 
> However, it needs to be mindful of case style (e.g CollectorBuilder should be transformed into collector and builder and check separately).

You just contradicted yourself, proving my point.  First you said that would be great behavior, but then you added a caveat, showing that you don't actually want simple comparison of tokens to dictionary entries.  You actually want the spell checker to be "smart" enough to split compound class names into separate tokens--basically you want it to be aware of differences that are a result of the fact that you're editing ***code*** and not normal English text.  You can't simultaneously claim this is a simple problem and also say that you want such benefits which require added complexity.  The two claims are contradictory.  You can't have it both ways.

Maybe it would be better to be thankful for all the work that the Zed programmers are doing for us, instead of complaining that it's not good enough.

**Comment by phaynes** [comment]
Dear @florinpatrascu,

Thank you very much for your response and for answering my questions. I found your insights very interesting and appreciate your considered post above.

Perhaps Zed's performant nature is the reason for this discussion. It is addictive, like driving a Ferrari. The choice to drive a Ferrari comes with emotion and passion, and of course, a price tag. You often have to accept that the Ferrari omits many features available in consumer cars, like a Ford.

The "driving pleasure" of Zed is something I haven't experienced since the '80s with my x86 assembly language scientific editor. Since then, due to economics, a reasonable IDE, and Word, is what I have had to use. Now, later in life, I routinely apply the [Marie Kondo principle of "Ask Yourself If It Sparks Joy"](https://konmari.com/marie-kondo-rules-of-tidying-sparks-joy/) when deciding whether to keep something. Reflecting on my over 25 years of Word usage, I find that Word does not "spark joy" for me. I am quite motivated to place it into the garbage bin and now have a clear and achievable pathway for doing so.

Florin, I completely respect your choice not to have an LLM running in parallel with your editor. You enjoy the Ferrari driving experience, accept it without heated seats, and are able to commute and live around your home without expecting to go cross-country — awesome. It is fascinating to see how different people live and the choices they make when you travel.

People are often a product of their experiences. Part of mine involves seeing how fantastically expensive it can become when writing precise programming tools, and that in practice, it is almost always harder than initially thought or reasonable. So, congratulations to the Zed team for the amazing work they have done. My ambition and budget are more modest. I aim to do all my programming and editing work in a single performant environment, with Word and its ecosystem, along with tools such as Grammarly, becoming just a painful memory.

In this context, I find connecting Zed to LLMs as definitely the lesser of two evils. My usage requires high-quality grammar checking, connection of different tools and security standards into the programming environment, and production of publication-grade documentation.

This trade-off is not for everyone. That stated, mostly I get to drive the Ferrari, but I accept that occasionally the Alfa is driven when ferrying the kids about. But at least the Ford can be sold.

Do look after yourself, and enjoy the Olympics. 

Philip

**Comment by bbb651** [comment]
In my opinion, an LLM based spell checking tool is an entirely different feature from a local spell checker, and arguing it is a good replacement is like saying a typewriter is a replacement for a pen - sure for some people for some tasks it's a much better tool, but for others it's simply not a replacement.

An LLM based spell checking tool currently has some very big drawbacks:
- Requires running a local LLM (on most computers today it requires large power draw, heat and noise) or relying on (and maybe paying for) some external server farm
- Significantly worst performance on languages that are not English
- Sometimes makes mistakes 

All of these will improve over time, and maybe at some point in the future would not make sense to have dictionary based spell checking, but *currently* this is not an alternative for the majority of people.

Also, that is not to say that language models cannot be used within such a local spell checker, but they have to be **tiny** enough to run on most configurations, **optional**, and still use a **dictionary as a source of truth**. One really good example of that is ordering of possible corrections - as a non-native English speaker I personally find myself a lot of times chucking a really badly spelled word into a search engine and it gives much better results than the builtin autocomplete in browsers. [Firefox had success doing something similar with a gpt2 based model for alt text generation](https://hacks.mozilla.org/2024/05/experimenting-with-local-alt-text-generation-in-firefox-nightly/).

**Comment by Retkid** [comment]
> Sometimes makes mistakes

All spell checkers make mistakes, human languages are not materially self consistent.

>  Significantly worst performance on languages that are not English

You cant say this about all, we need to talk about a specific implementation. A better argument is lesser language availability, we you need a lot of information to train them on, and only major languages have this. I'm sure pre-existing models have good performance for English, Japanese, French, Spanish, Italian, German, Russian, Chinese, Korean, Arabic, but less so for Urdu, Ukrainian, Swedish, Norwegian,  Icelandic, Swahili, Finnish, etc, etc, etc. Data is not an issue for spell checkers written by native speakers.


> Requires running a local LLM (on most computers today it requires large power draw, heat and noise) or relying on (and maybe paying for) some external server farm

Big issue, because, you need large models for each language you implement, how do we store and distribute these? Are we expecting everyone using Zed to have a g pu? To have a cpu? To have the spare memory for these models? This is an absurd imp,concept, when there are per-existing traditional spell checkers which arre good enough.

Anyway, lastly. We don't need something that much superior to VSCODE, we need not be determined to prove ourselves with superfluous, flashy features. We should strive for functionality,  and maintainability first and foremost.

**Comment by Conaclos** [comment]
> My dream would be to embed something like [Grammarly](https://app.grammarly.com/) right in Zed :)

I personally use Language Tools because you can deploy your own instance locally.
It could be great to have this kind of integration with zed.

**Comment by jansol** [comment]
There seems to be a LSP implementation for LanguageTool. Getting that to work with Zed should be as easy as copying some existing language extension and pointing it to that LSP.

**Comment by ivansigmund** [comment]
> There seems to be a LSP implementation for LanguageTool. Getting that to work with Zed should be as easy as copying some existing language extension and pointing it to that LSP.

How could i do that, i am a noob. Thank you for any help

**Comment by bbb651** [comment]
There are is a bit of documentation [here](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md) on extensions, read it, look at a couple of random extensions with an LSP in that repository, specifically the rust code they contain, most of them should be pretty similar with code downloading the LSP from github. Copy the code and change the relevant strings, copy an `extension.toml`, change it as needed and get rid of `language` since it isn't needed here. You can test it with "load dev extension"

**Comment by ivansigmund** [comment]
> There are is a bit of documentation [here](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md) on extensions, read it, look at a couple of random extensions with an LSP in that repository, specifically the rust code they contain, most of them should be pretty similar with code downloading the LSP from github. Copy the code and change the relevant strings, copy an `extension.toml`, change it as needed and get rid of `language` since it isn't needed here. You can test it with "load dev extension"

Thank you very much for your answer. I did try an installated also my extension, however i do not understand what would be the relevant parameters to define how to show wrongly spelled words etc. 
If someone managed to do this it would be really helpful to see their code (if someone is willing to share it)

**Comment by notpeter** [comment]
I've created an dedicated issue on the extensions repo for discussion of a LanguageTool based Spell-checker extensions.  Please migrate your discussion there as this issue has has 20+ followers who get notifications for every comment.

- See: https://github.com/zed-industries/extensions/issues/1332

Thanks @ivansigmund @bbb651 @jansol @Conaclos for identifying this a potential path forward.

**Comment by camstuart** [comment]
+1 [cspell](https://cspell.org/) would be my vote!

**Comment by notpeter** [comment]
> +1 [cspell](https://cspell.org/) would be my vote! 

I've created a dedicated issue on the extensions repo for cspell via LSP:

https://github.com/zed-industries/extensions/issues/1414

**Comment by JosephTLyons** [comment]
Just popping in to say that someone has contributed a `typos` extension.

<img width="1468" alt="SCR-20240920-nofy" src="https://github.com/user-attachments/assets/707b900f-6240-4994-aa83-ce04db0de186">

**Comment by mocenigo** [comment]
> There seems to be a LSP implementation for LanguageTool. Getting that to work with Zed should be as easy as copying some existing language extension and pointing it to that LSP.

Ideally this should work together with some specific LSP.  For instance, in C you use some C specific language server to identify problems in the programming, but for comments and strings use a spell/grammar checker.

For markdown and LaTeX the ability of using two systems is even more important. I do not understand why this is not yet properly supported.

**Comment by rucoder** [comment]
> Just popping in to say that someone has contributed a `typos` extension. I think we should leave this open though, as the underlying [typos](https://crates.io/crates/typos) crate is intentionally low-false positive, so it doesn't catch as much as a normal spell checker.
> <img alt="SCR-20240920-nofy" width="1468" src="https://private-user-images.githubusercontent.com/19867440/369530505-707b900f-6240-4994-aa83-ce04db0de186.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyNjY2NzIsIm5iZiI6MTcyOTI2NjM3MiwicGF0aCI6Ii8xOTg2NzQ0MC8zNjk1MzA1MDUtNzA3YjkwMGYtNjI0MC00OTk0LWFhODMtY2UwNGRiMGRlMTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDE1NDYxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyZmI0YjZhMWU2ZTQxOWY0YWU2ZDEzNjk3OTg0YTViZmZiYjc3YWFkMWVkOWYyODU3MGFiOTcwMjIyYzlkNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cGZzWO7BTDfRmsqQhO3QWLM2ya9GsX5tn2IC8kTGDLo">
> 
> See: [zed-typos extension](https://github.com/BaptisteRoseau/zed-typos.git) for more.

it works pretty good for me, but I cannot find a way to 'ignore' a word. Is there any configuration file for typos extension?

**Comment by bbb651** [comment]
Yes, see the [typos reference](https://github.com/crate-ci/typos/blob/master/docs/reference.md) and the [README](https://github.com/crate-ci/typos#false-positives).
typos-lsp has [an open issue](https://github.com/tekumara/typos-lsp/issues/12) to add an lsp action for it, and there's [a draft PR](https://github.com/tekumara/typos-lsp/pull/72) to implement it.

**Comment by kakalot0008** [comment]
I installed the `typos` extension, but it's not working. How can I configure this?

**Comment by kakalot0008** [comment]
I installed the `typos` extension, but it's not working. How can I configure it?

**Comment by kakalot0008** [comment]
Hi, @scorphus. It's worked. Thanks for your setting.

**Comment by blopker** [comment]
Hey all! I took some time to review all the solutions presented so far and wrote up a survey doc: https://gist.github.com/blopker/2a56b205eaaeb4f10ed7a4c2729c27c2 Would love any feedback people have.

The TLDR is that I think there are two ways to go here:

1. Create a language server around CSpell. This has downsides though (in the above doc).
2. Or create a code-aware language server around Helix's [Spellbook](https://github.com/helix-editor/spellbook). This would be more work, but I think a better outcome.

Thoughts?

**Comment by mocenigo** [comment]
> I would prefer LanguageTool to Grammarly as it FOSS and supports more languages

No, please. It is very slow and uses a LOT of resource as I tried using ltex — that uses LanguageTool as a backend – with Zed and I used it also with VSCode. The experience was extremely frustrating.

**Comment by mocenigo** [comment]
> I installed the `typos` extension, but it's not working. How can I configure it?

It corrects very few errors — because the developers want very few false positives. And only with some file types.

**Comment by mocenigo** [comment]
> Hey all! I took some time to review all the solutions presented so far and wrote up a survey doc: https://gist.github.com/blopker/2a56b205eaaeb4f10ed7a4c2729c27c2 Would love any feedback people have.
> 
> The TLDR is that I think there are two ways to go here:
> 
> 1. Create a language server around CSpell. This has downsides though (in the above doc).
> 2. Or create a code-aware language server around Helix's [Spellbook](https://github.com/helix-editor/spellbook). This would be more work, but I think a better outcome.
> 
> Thoughts?

I believe that Zed should have built in support for spell checking, not grammar. I think grammar, as well as style, should be a matter for more specialised tools, like the inline AI assistants. 

This support should be integrated because the system must first recognise whether some character triggers the operation of a language server, and not interfere with the latter as long as it has control. 

Since “spell checking” is part of the old 2024 timeline set by @nathansobo , I believe the plan looks like something along these lines.

**Comment by ygingras** [comment]
I tried the ltex extension for Zed and it does a really good job of figuring out that comments should be grammatically correct text while variable names should be only be spelled mostly correctly. It also does this for a bunch of programming languages. I'm not sure how it works, but I like that part a lot. I can also supply a custom dictionary on a per programming language basis. There is a pop-up to add words to my dictionary when I right click on an underlined word in Zed, but that does not work for me and the word stays underlined until I add the word to my dictionary file manually. 

I would not want my code to be sent to a third party. Having the option to run spell checking completely local should be a pretty high priority. 

I regularly write in french and in english, sometimes in the same document when I make a bilingually document, let's say a draft of a Meetup event, which does not support language tagging. [example](https://www.meetup.com/montreal-python/events/305050771/?eventOrigin=group_events_list)

Being able to quickly switch the spell checker from one language to the other is pretty important to me. Even better is Firefox that allows me to enable a bunch of dictionaries all at the same time. This way I don't have to switch, at the very small risk that I will introduce french words in the english parts of my document. I'm fine with that.

**Comment by phaynes** [comment]
> Hey all! I took some time to review all the solutions presented so far and wrote up a survey doc: https://gist.github.com/blopker/2a56b205eaaeb4f10ed7a4c2729c27c2 Would love any feedback people have.
> 
> The TLDR is that I think there are two ways to go here:
> 
> 1. Create a language server around CSpell. This has downsides though (in the above doc).
> 2. Or create a code-aware language server around Helix's [Spellbook](https://github.com/helix-editor/spellbook). This would be more work, but I think a better outcome.

Hi Bo,

Thank you for your considered analysis.

I think an underlying issue when it comes to code spell-checking is that there are probably several use cases that all fall under the banner of "code spell-checking".

**Spell Checking as Auto Correct**: There is definitely the use case where people want a fast inline spell-checker; for some, it is as an aid to typing.

However, as I touch type, any automatic inline spell-checking gets in the way, and thus performing spell check as a discrete step is perfectly fine.

Furthermore, I see the narrow scoping of spell check as problematic, since spelling is often associated with grammar (e.g., their/there, principle/principal). Being Australian, I am quite fussy about International English spelling (e.g., *colour* not *color*). While I produce all my written work in Zed, when proofing I find the ability to "choose your style" in Claude.ai as very helpful.

Zed already deals with the issue of AI keys and is able to use both local and remote LLMs. So while I accept your analysis of locality for "spellchecking as autocorrect", I don't see that it holds generally.

Furthermore, given that the coding ability of AIs is rapidly approaching science fiction levels—prompting, specification, and other high-level language-based tasks—I am struggling to find use cases where my development work is not connected to AIs, particularly in larger software engineering teams.

But on the other hand, this is not how everyone will work.

Philip

**Comment by jansol** [comment]
Daydreaming about the capabilities of a fictional AI is nice and all, but many organizations do in fact care about their data and explicitly prohibit sending it overseas for black box processing, at the very least in order to comply with local regulations such as the GDPR.

This means that if Zed has non-local spell checking, it does effectively not have spell checking. And if said spell checking can't be completely disabled, Zed can't be used at such orgs at all.

**Comment by brandondrew** [comment]
> The TLDR is that I think there are two ways to go here:
> 
> 1. Create a language server around CSpell. This has downsides though (in the above doc).
> 2. Or create a code-aware language server around Helix's [Spellbook](https://github.com/helix-editor/spellbook). This would be more work, but I think a better outcome.
> 
> Thoughts?

Wow, thanks for all the work you put into researching that!

Here's a summary that is slightly more detailed than your TLDR, but less detailed than the full gist:


- There doesn't currently seem to be a portable code spell checker that would work for most people. All the current solutions either are editor-specific, do too much, or do too little.
- Different people have different goals for the spell-checking, so there are open questions around scope, locality, languages, use of native APIs, libraries, configuration, and UI.
- It seems best that a code spell checker not try to fix grammar issues, and that it not try to also cover prose.
- Both for privacy and performance it seems best for the spell checker to keep all data local, and therefore to run locally.
- It seems best to prioritize flexibility in programming languages, although supporting multiple natural languages should also be a goal.
- Native spell checkers in operating systems are not necessarily well-suited for programming languages.

- There don't seem to be any current solutions that match our needs.

  + [Code Spell Checker for VSCode](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) — no language server, thus not usable by Zed; written in Typescript thus not as performant as we might want;
  + [Harper](https://github.com/Automattic/harper/tree/master) — explicitly focused on English prose, not code in multiple languages;
  + [Language Tool](https://languagetool.org/) — written in Java and memory-hungry;
  + [Vale](https://github.com/errata-ai/vale) — more of a linter than a spell-checker; the complicated configuration that is required creates an unnecessary barrier to entry;
  + [Typos](https://github.com/crate-ci/typos) — has a ton of false negatives, catching only a few errors because it's meant to be run in CI; uses a hardcoded 'typos' word list, creating a maintenance burden around supporting multiple natural languages;
  + [Spellbook](https://github.com/helix-editor/spellbook) — not a complete solution, but could be a powerful building block; written in Rust, a Hunspell-compatible spell checking library, if someone writes a language server for it.

Best options:

1. write a language server for CSpell: fastest route to a solution, but it comes with the JavaScript performance overhead;
2. build a code spell checker around Spellbook. This would involve a bit more work since a language server needs to be created, along with working out the complexities around parsing code into words to feed into it.

**Comment by jansol** [comment]
> 2. build a code spell checker around Spellbook. This would involve a bit more work since a language server needs to be created, along with working out the complexities around parsing code into words to feed into it.

I like this option -- I'd argue that it should not be done via a language server (and thus not as a plugin, at least not without adding some more API for plugins).

A plain dictionary-based spell checker could easily be hooked into the existing tree-sitter integration which gives it knowledge about keywords etc that need to be excluded, essentially for free. This also means it can check keywords against an English or other programming language-specific dictionary regardless of what natural language is used otherwise. And adding support for shipping *dictionaries* as plugins should be quite uncontroversial, especially if they use a well-known format to begin with.

I think a major requirement for this path would be having a context menu item for adding a new word to a local user dictionary as well as a view for managing said user dictionary (add/remove/edit words). The requirement for custom UI means again that this can't be done as an extension for the foreseeable time.

**Comment by brandondrew** [comment]
> I think a major requirement for this path would be having a context menu item for adding a new word to a local user dictionary as well as a view for managing said user dictionary (add/remove/edit words). The requirement for custom UI means again that this can't be done as an extension for the foreseeable time.

While I agree a context menu would be good to have, the other parts of this work could proceed without waiting until we can create a context menu.  We're already accustomed to editing config files for Zed without any UI around our preferences and settings.  We can do the same thing with word lists until the time when it is possible to add a context menu.

**Comment by phaynes** [comment]
> Daydreaming about the capabilities of a fictional AI is nice and all, but many organizations do in fact care about their data and explicitly prohibit sending it overseas for black box processing, at the very least in order to comply with local regulations such as the GDPR.
> 
> This means that if Zed has non-local spell checking, it does effectively not have spell checking. And if said spell checking can't be completely disabled, Zed can't be used at such orgs at all.

Mr Solanti, framing “non-local spell checking” as the end of Zed adoption overlooks the existence of locally hosted AI solutions like Ollama. GDPR, while important, is hardly an unsolvable problem.

With AI tools boasting competitive programming ELOs above 2,000 soon becoming the norm, engineers must adapt—or explore new career options. Embedding an AI pair programmer in Zed could elegantly handle tasks like spell checking while maintaining compliance.

Let’s focus on exploring all possibilities to make Zed an even greater tool for a wider audience.

**Comment by mocenigo** [comment]
@phaynes I am not sure most people would be comfortable with an editor that uses a few GB of RAM to spell check. Let us focus on the actual problem of spell checking, lightweight and using little memory — fancier AI based systems can always be implemented as add-ons, ie extensions. And be optional.

**Comment by jansol** [comment]
> While I agree a context menu would be good to have, the other parts of this work could proceed without waiting until we can create a context menu.

Yes.
* [ ] loading dictionaries
* [ ] checking non-keyword tree-sitter nodes against the dictionary (or only specific captures? i.e. identifiers and/or string literals)
* [ ] attaching "diagnostics" to misspelled nodes/words

are things that could be worked on right away by anyone. Extensions don't have access to the tree-sitter data so it would have to happen in the main Zed code base though.

Then once the basic functionality is there we can look into the UI side. And the Zed team will probably want to do their own UI design pass on it at that point, to keep things cohesive and in line with their vision.

**Comment by blopker** [comment]
I would like to see a spell checker built into Zed. However, I'm also in favor of an extension, especially if it helps us get something working sooner. I'm often having to copy text around for spell checking, it's a productivity drain. An extension is also something I could bring into any LS capable editor, which I think would be huge for a lot of people.

I put a _rough_ POC together that attempts to mix tree-sitter + word splitting heuristics + Spellbook into the beginnings of an LS: https://github.com/blopker/codecheckr. If someone is up to help me with this, let me know, but I'll need a lot of help with my Rust 😄 

I found that tree-sitter is fast. Reusing Zed's data may not be that beneficial. Plus, having control over the tree-sitter queries is nice. [Example here](https://github.com/blopker/codecheckr/blob/main/spellbook-lsp/src/queries.rs).

**Comment by brandondrew** [comment]
> I'm often having to copy text around for spell checking, it's a productivity drain.

I'm not trying to dispute any of the points you're making, but while we wait for better support in Zed for checking spelling, you can have a file open in multiple editors at the same time.  While I (obviously) haven't tried every single pair of editors, I've tried several and rarely (if ever) had problems.  So you can do spell checking in XYZ Editor while you do most of your work in Zed.  Not ideal, but much better than copying and pasting between apps.

**Comment by mocenigo** [comment]
> I found that tree-sitter is fast. Reusing Zed's data may not be that beneficial. Plus, having control over the tree-sitter queries is nice. [Example here](https://github.com/blopker/codecheckr/blob/main/spellbook-lsp/src/queries.rs).

About this "reusing" — so there could be more copies of the tree-sitter data and tree-sitter internal structures for the same file? Since in the past for some cases the data structures of tree-sitter were exploding, I am not sure this is a good idea.

**Comment by blopker** [comment]
I've been hacking away on the language server/extension path, and I'm at the stage now where I was able to fix a misspelling in my own README for this extension 😅. 

<img width="943" alt="image" src="https://github.com/user-attachments/assets/cb14f17d-36da-4de7-aa57-5469f3e4b91d" />

Oops.

It's still a ways away from being usable though. The extension doesn't have any quick actions, just the message with suggestions in it. It also is a bit too aggressive with URLs. I think we'll need a dictionary of common jargon terms to ignore as well.

> Since in the past for some cases the data structures of tree-sitter were exploding, I am not sure this is a good idea.

I'd guess many extensions are doing their own parsing? I'm not sure what the alternative is besides using Zed's structures, but I don't think there's a way for an extension to do that. I do think there are ways to minimize the impact in this case, like only spell checking on open and save.

**Comment by jansol** [comment]
Extensions don't do any parsing themselves, they merely ship a tree-sitter grammar for Zed to load. Extensions can also download and run language servers, and those of course parse the code... but they typically rely on a full-on compiler/interpreter for the language in question for that, not tree-sitter.

**Comment by martinhjartmyr** [comment]
I just tried [zed-cspell](https://github.com/mantou132/zed-cspell) and it seems to be working very well. I'm very happy with `cspell`, been using it in neovim for years.

**Comment by mocenigo** [comment]
> I just tried [zed-cspell](https://github.com/mantou132/zed-cspell) and it seems to be working very well. I'm very happy with `cspell`, been using it in neovim for years.

It seems to be a very recent addition, and the big improvement wrt ltex and language tool based solutions is that it is fast. 

I like it!

Is there a way to have it also suggest completions?

**Comment by mocenigo** [comment]
> Ok, update on [Codebook](https://github.com/blopker/codebook)! I'm very happy with it so far, and I'd love any feedback.

Even before downloading it: glorious!


> See [supported languages](https://github.com/blopker/codebook?tab=readme-ov-file#supported-languages), and please let me know what should be added next! 🤠

LaTeX (TeX in general) and Bibtex!

**Comment by JosephTLyons** [comment]
Hey all, it seems like we have a few choices of spell check here, through extensions, so I think we can close this issue out:

- `[codebook](https://github.com/blopker/codebook/tree/main?tab=readme-ov-file)`
- `[harper](https://github.com/Stef16Robbe/harper_zed)`
- `[typos](https://github.com/BaptisteRoseau/zed-typos)`

Each one, available through the extension store.

If you run into issues, please open issues in the extension's repository! Many thanks to the extension authors for closing this out. ❤️

**Comment by jvoisin** [comment]
As those are LSP, would it be possible to make us of them in Zed without having to install extensions?

**Comment by nixpulvis** [comment]
Anyone know of a spell check package which easily allows you to toggle it on and off? I tried a few but didn't find a way.

EDIT: Or how to configure the typos LSP, I didn't see it in the docs when I looked.

**Comment by blopker** [comment]
Hey, I'm the author of Codebook. I don't know how Zed would actually handle toggling LSPs on and off, but I'm open to adding it if there's a way.
